pkg/services/sqlstore/migrations/ualert/template_test.go

```diff
@@ -0,0 +1,321 @@
+		offset: 2,
+		err:    "ambiguous delimiter",
+	}, {
+		name:   "variable without closing brace returns error",
+		text:   "${instance is down",
+		offset: 18,
+		err:    "expected '}', got 'n'",
+	}}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			token, offset, err := tokenizeVariable([]rune(test.text))
+			if test.err != "" {
+				assert.EqualError(t, err, test.err)
+			}
+			assert.Equal(t, test.offset, offset)
+			assert.Equal(t, test.token, token)
+		})
+	}
+}
+
+func TestTokenizeTmpl(t *testing.T) {
+	tests := []struct {
+		name   string
+		tmpl   string
+		tokens []Token
+	}{{
+		name:   "simple template can be tokenized",
+		tmpl:   "${instance} is down",
+		tokens: []Token{{Variable: "instance"}, {Literal: " is down"}},
+	}, {
+		name: "complex template can be tokenized",
+		tmpl: "More than ${value} ${status_code} in the last 5 minutes",
+		tokens: []Token{
+			{Literal: "More than "},
+			{Variable: "value"},
+			{Literal: " "},
+			{Variable: "status_code"},
+			{Literal: " in the last 5 minutes"},
+		},
+	}, {
+		name:   "variables without spaces between can be tokenized",
+		tmpl:   "${value}${status_code}",
+		tokens: []Token{{Variable: "value"}, {Variable: "status_code"}},
+	}, {
+		name:   "variables without spaces between then literal can be tokenized",
+		tmpl:   "${value}${status_code} in the last 5 minutes",
+		tokens: []Token{{Variable: "value"}, {Variable: "status_code"}, {Literal: " in the last 5 minutes"}},
+	}, {
+		name: "variables with reserved characters can be tokenized",
+		tmpl: "More than ${$value} ${{status_code} in the last 5 minutes",
+		tokens: []Token{
+			{Literal: "More than "},
+			{Variable: "$value"},
+			{Literal: " "},
+			{Variable: "{status_code"},
+			{Literal: " in the last 5 minutes"},
+		},
+	}, {
+		name:   "ambiguous delimiters are tokenized as literals",
+		tmpl:   "Instance ${instance and ${instance} is down",
+		tokens: []Token{{Literal: "Instance ${instance and "}, {Variable: "instance"}, {Literal: " is down"}},
+	}, {
+		name:   "all '$' runes preceding a variable are included in literal",
+		tmpl:   "Instance $${instance} is down",
+		tokens: []Token{{Literal: "Instance $"}, {Variable: "instance"}, {Literal: " is down"}},
+	}, {
+		name:   "sole '$' rune is included in literal",
+		tmpl:   "Instance $instance and ${instance} is down",
+		tokens: []Token{{Literal: "Instance $instance and "}, {Variable: "instance"}, {Literal: " is down"}},
+	}, {
+		name:   "extra closing brace is included in literal",
+		tmpl:   "Instance ${instance}} and ${instance} is down",
+		tokens: []Token{{Literal: "Instance "}, {Variable: "instance"}, {Literal: "} and "}, {Variable: "instance"}, {Literal: " is down"}},
+	}, {
+		name:   "variable with newline tokenized as literal",
+		tmpl:   "${value}${status_code\n}${value} in the last 5 minutes",
+		tokens: []Token{{Variable: "value"}, {Literal: "${status_code\n}"}, {Variable: "value"}, {Literal: " in the last 5 minutes"}},
```

Comment:
This contradicts https://github.com/grafana/grafana/pull/62911/files#diff-c7f0bb3b124090c2c3acef55c5a6ae1df79fead72e9fb7c3f3626ce59284f020R65-R70. I don't understand how it works here and does not there.